src/drivers/pwm_out_sim/PWMSim.cpp

```diff
@@ -0,0 +1,666 @@
+				}
+
+				_mixers->add_mixer(mixer);
+				_mixers->groups_required(_groups_required);
+			}
+
+			break;
+		}
+
+	case MIXERIOCLOADBUF: {
+			const char *buf = (const char *)arg;
+			unsigned buflen = strnlen(buf, 1024);
+
+			if (_mixers == nullptr) {
+				_mixers = new MixerGroup(control_callback, (uintptr_t)&_controls);
+			}
+
+			if (_mixers == nullptr) {
+				_groups_required = 0;
+				ret = -ENOMEM;
+
+			} else {
+
+				ret = _mixers->load_from_buf(buf, buflen);
+
+				if (ret != 0) {
+					PX4_ERR("mixer load failed with %d", ret);
+					delete _mixers;
+					_mixers = nullptr;
+					_groups_required = 0;
+					ret = -EINVAL;
+
+				} else {
+					_mixers->groups_required(_groups_required);
+				}
+			}
+
+			break;
+		}
+
+
+	default:
+		ret = -ENOTTY;
+		break;
+	}
+
+	unlock();
+
+	return ret;
+}
+
+int
+PWMSim::task_spawn(int argc, char *argv[])
+{
+	_task_id = px4_task_spawn_cmd("pwmsim",
+				      SCHED_DEFAULT,
+				      SCHED_PRIORITY_ACTUATOR_OUTPUTS,
+				      1000,
+				      (px4_main_t)&run_trampoline,
+				      nullptr);
+
+	if (_task_id < 0) {
+		_task_id = -1;
+		return -errno;
+	}
+
+	// wait until task is up & running (the mode_* commands depend on it)
+	if (wait_until_running() < 0) {
+		_task_id = -1;
+		return -1;
+	}
+
+	return PX4_OK;
+}
+
+PWMSim *PWMSim::instantiate(int argc, char *argv[])
+{
+	return new PWMSim();
+}
+
+int PWMSim::custom_command(int argc, char *argv[])
+{
+	const char *verb = argv[0];
+
+	/* start the FMU if not running */
+	if (!is_running()) {
+		int ret = PWMSim::task_spawn(argc, argv);
+
+		if (ret) {
+			return ret;
+		}
+	}
+
+	Mode servo_mode = MODE_NONE;
+
+	if (!strcmp(verb, "mode_pwm")) {
+		servo_mode = PWMSim::MODE_8PWM;
+
+	} else if (!strcmp(verb, "mode_pwm16")) {
+		servo_mode = PWMSim::MODE_16PWM;
+	}
+
+	/* was a new mode set? */
+	if (servo_mode != MODE_NONE) {
+		/* switch modes */
+		PWMSim *object = get_instance();
+
+		if (servo_mode != object->get_mode()) {
+			/* (re)set the PWM output mode */
+			return object->set_mode(servo_mode);
+		}
+	}
+
+	return print_usage("unknown command");
+}
+
+int PWMSim::print_usage(const char *reason)
+{
+	if (reason) {
+		PX4_WARN("%s\n", reason);
+	}
+
+	PRINT_MODULE_DESCRIPTION(
+		R"DESCR_STR(
+### Description
+
+Driver/configurator for the virtual PWMSim port.
+
+This virtual driver emulates PWM / servo outputs for setups where
+the connected hardware does not provide enough or no PWM outputs.
+
+Its only function is to take actuator_control uORB messages,
+mix them with any loaded mixer and output the result to the
+actuator_output uORB topic. PWMSim can also be performed with normal
+PWM outputs, a special flag prevents the outputs to be operated
+during PWMSim mode. If PWMSim is not performed with a standalone FMU,
+but in a real system, it is NOT recommended to use this virtual
+driver. Use instead the normal FMU or IO driver.
+
+)DESCR_STR");
+
+	PRINT_MODULE_USAGE_NAME("pwmsim", "driver");
```

Comment:
This needs to match the shell command: `pwm_out_sim` (and the same for `px4_task_spawn_cmd`).